### ISSUE TITLE
[Question] Show metadata on BubbleUpnp

### ISSUE THREAD

**Issue by GioF71** [philippe44/SpotConnect#31]
Hello, today I restarted my spotify account, and tried your SpotConnect again, but I wanted to see if an instance of BubbleUpnp can act as a display for what's playing on SpotConnect.
Unfortunately, this does not happen for me. In some cases I see album art from a previous track, I cannot seem to be able to follow what's playing through BubbleUpnp.
Do you have any suggestion?
Thank you very much!

**Comment by GioF71** [comment]
If I select the OpenHome renderer in BubbleUpnp, I get something, but there is no album art and the metadata is not correct.

![image](https://github.com/philippe44/SpotConnect/assets/5603160/3e2cb8ed-06f7-4409-ae75-5318ee8500ee)

I am available to any test you would like me to do if this can be helpful.
Thank you anyway for your great software.

**Comment by philippe44** [comment]
The TrackNumber and DiscNumber were not provided by CSpot, I've added them and done a PR. Duration is set

**Comment by philippe44** [comment]
It's in 0.8.0. I don't know why duration does not appear. I can't easily test BubbleUPnP as I need to fire another VM or machine where I'll install Android Hyper-V crap. I did that on my main machine and it killed my other VM system, it took me 5 hours to roll back everything properly (it did a lot of damage, I'm usually more careful).

**Comment by GioF71** [comment]
I am really sorry that this caused such a damage to you. I will do the tests on BubbleUpnp when needed, just let me know if this need arise I will be happy to help you :-)
I will try version 0.8.0 as soon as I can, thank you!

**Comment by philippe44** [comment]
Not your fault at all, completely mine. I've been lazy and did not do my normal safeguard process

**Comment by philippe44** [comment]
For the artwork, it should be retrievable from the player. Just use a tool like UPnP DeviceSpy (or anyother) to query the "GetPositionInfo" of your device.

**Comment by GioF71** [comment]
Ok. there is a lot of progress with version 0.8.0.

If I connect BubbleUpnp to the non-openhome renderer, and the playlist is clear, when I try to play from spotify, everything works. Also the coverart.

But:

If I seek the track, the indicator restarts from the beginning.
When the track finishes, so there is a track change, the metadata don't change (art, title, album, length).
Also if I pick another song and want to play it, it plays correctly, but again the metadata don't change (art, title, album, length).
Instead, if I hit pause before selecting a new track, the information about the new track (coverart included) is then displayed correctly.

If there are some tracks in the playlist (a playlist probably owned by the control point), if I hit play on spotify, that track plays, but the control point "thinks" it is playing the first track that was in that playlist. I suppose that, on the playback event, maybe the existing playlist could be cleared?

Maybe also on song change, if the playlist was cleared, the issue with the previous metadata being shown could be avoided.

If, instead, I select the openhome renderer in BubbleUpnp, the behaviour is a lot more consistent. The metadata that is available is always correct, but the coverart and track length are always missing in this scenario. 

Does this help?
If you want I can produce a video showing the spotify app and the bubbleupnp app (captured using qtscrcpy), let me know if this can be helpful.

Thank you!

**Comment by GioF71** [comment]
p.s. I can't find the UPnP devicespy, if [this](https://www.meshcommander.com/upnptools) is the site, the binaries are not available anymore. I will look for some alternatives

**Comment by philippe44** [comment]
Yes a screen live capture would be helpful as it's difficult to follow. One thing though: restarting from 0 on the UPnP device when seeking is unavoidable.

**Comment by philippe44** [comment]
Another thing about the lack of detection of change: uPnP is not a great protocol, to say it politely, and implementation by many devices is abysmal. When you play another track, there is a quick stop then play set of state changes and I suspect bubbleupnp does not capture these. Normally, uPnP has an eventing mechanism to make that unmissable but most don't work and if the application does not do polling, the you miss state changes and so here bubbleupnp would have missed the fact that the track changed. Polling is bad as well if you just do state polling as you miss the stop in play,stop,play set of status. You need to poll the track info as well. Most of my code is 1 garbage collection of special cases and timers because of all that crap.

**Comment by GioF71** [comment]
Understood! I totally believe you.
All things considered, BubbleUpnp as a display works very well when using another device as the real control point, again with bubbleupnp and playing from other media servers, or maybe from BubbleUpnp's Qobuz and Tidal implementations/proxies.
When I want to play from Spotify (which I currently don't do on a regular basis) it might be acceptable to switch the "display" tablet from BubbleUpnp to the Spotify app so it would show what's playing. 
Anyway I will try to prepare you a video as soon as possible. I just made some test, I only need to follow a script in order to keep the video short.

**Comment by GioF71** [comment]
Hello, quick update!
After discovering that upplay (from the author of upmpdcli) always correctly displays the correct metadata, even if I do not hit pause before a track change, I enabled "Use eventing" on bubbleupnp on the relevant renderer, and now BubbleUpnp also works properly!!

**Comment by GioF71** [comment]
Ops, it just stopped working after a moment... so confusing!

**Comment by philippe44** [comment]
> Ops, it just stopped working after a moment... so confusing!
>
Not surprising and consistent with what uPnP is... some events are missed.

**Comment by GioF71** [comment]
I have written on the bubbleupnp forum [here](https://xdaforums.com/t/app-7-0-v4-3-1-2-bubbleupnp-upnp-dlna-chromecast-control-point-and-renderer.1118891/post-89218359), I will report here if there is a solution or a workaround.
